File: CPP/DP_Math/Fibonacci.cpp
/**
 * Fibonacci Sequence Implementation (Dynamic Programming)
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(1) iterative approach
 */

#include <iostream>
#include <vector>

long long fibonacci(int n) {
    if (n <= 1) return n;
    
    long long prev = 0, curr = 1;
    for (int i = 2; i <= n; i++) {
        long long next = prev + curr;
        prev = curr;
        curr = next;
    }
    return curr;
}

/*
int main() {
    std::cout << fibonacci(10) << std::endl; // Output: 55
    return 0;
}
*/

File: TypeScript/DP_Math/Fibonacci.ts
/**
 * Fibonacci Sequence Implementation (Dynamic Programming)
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(n) with memoization array, or O(1) with iterative approach.
 */

export class Fibonacci {
    /**
     * Recursive approach with Memoization
     */
    static getWithMemo(n: number, memo: Map<number, number> = new Map()): number {
        if (n <= 1) return n;
        if (memo.has(n)) return memo.get(n)!;

        const result = this.getWithMemo(n - 1, memo) + this.getWithMemo(n - 2, memo);
        memo.set(n, result);
        return result;
    }

    /**
     * Iterative approach (Space Optimized)
     */
    static getIterative(n: number): number {
        if (n <= 1) return n;
        
        let prev = 0;
        let curr = 1;

        for (let i = 2; i <= n; i++) {
            const next = prev + curr;
            prev = curr;
            curr = next;
        }

        return curr;
    }
}

// Example usage:
// console.log(Fibonacci.getIterative(10)); // Output: 55
